orca-core/src/test/groovy/com/netflix/spinnaker/orca/pipeline/util/PackageInfoSpec.groovy

```diff
@@ -16,19 +16,135 @@
+import com.netflix.spinnaker.orca.pipeline.StageDefinitionBuilder
 import com.netflix.spinnaker.orca.pipeline.model.Pipeline
 import com.netflix.spinnaker.orca.pipeline.model.PipelineStage
 import com.netflix.spinnaker.orca.pipeline.model.Stage
+import com.netflix.spinnaker.orca.pipeline.model.TestPipelineStage
 import org.springframework.beans.factory.annotation.Autowired
 import spock.lang.Specification
 import spock.lang.Unroll
 
+import java.util.function.BiFunction
+
 class PackageInfoSpec extends Specification {
 
   @Autowired
   ObjectMapper mapper
 
+  def "If no artifacts in current build info or ancestor stages, code should execute as normal"() {
+    final Pipeline pipeline = Pipeline.builder().withTrigger(["buildInfo": ["artifacts": [["fileName": "testFileName"]]]]).build()
+    final Stage quipStage = new PipelineStage()
+    quipStage.execution = pipeline
+    quipStage.context = [buildInfo: [name: "someName"], package: "testPackageName"]
+
+    final PackageType packageType = PackageType.DEB
+    final PackageInfo packageInfo =
+      new PackageInfo(quipStage, packageType.packageType, packageType.versionDelimiter, true, false, new ObjectMapper())
+    def allowMissingPackageInstallation = false
+
+    when:
+    packageInfo.findTargetPackage(allowMissingPackageInstallation)
+
+    then:
+    final ex = thrown(IllegalStateException)
+    ex.message.startsWith("Unable to find deployable artifact starting with")
+  }
+
+  def "If no artifacts in current build info, find from single ancestor that does"() {
+
+    final Map context = [
+      buildInfo: [
+        artifacts: [
+          [fileName: "testPackageName_1.0abc298.deb"],
+          [fileName: "testPackageName_1.0abc298.war"],
+          [fileName: "build.properties"]
+        ],
+        url: "http://localhost",
+        name: "testBuildName",
+        number: "1"
+      ]
+    ]
+
+    final Pipeline pipeline = new Pipeline()
+    final Stage quipStage = new PipelineStage() {
+      @Override
+      List<StageNavigator.Result> ancestors(BiFunction<Stage<Pipeline>, StageDefinitionBuilder, Boolean> matcher) {
+        final stage = new TestPipelineStage()
+        stage.context << context
+        return [new StageNavigator.Result(stage, stage)]
+      }
+    }
+    quipStage.execution = pipeline
+    quipStage.context = [buildInfo: [name: "someName"], package: "testPackageName"]
+
+    final PackageType packageType = PackageType.DEB
+    final PackageInfo packageInfo =
+      new PackageInfo(quipStage, packageType.packageType, packageType.versionDelimiter, true, false, new ObjectMapper())
+
+    when:
+    final Map requestMap = packageInfo.findTargetPackage(false)
+
+    then:
+    final fileName = context.buildInfo.artifacts[0].fileName
+    requestMap.package == fileName.substring(0, fileName.indexOf('.deb'))
+  }
+
+  def "If no artifacts in current build info, find from first ancestor that does"() {
+    final Map goodContext = [
```

Comment:
Any opportunity to reduce the duplication and cut these tests down from ~50 lines each?